tests/lib/rules/no-array-prototype-extensions.js

```diff
@@ -0,0 +1,276 @@
+'use strict';
+
+const rule = require('../../../lib/rules/no-array-prototype-extensions');
+const RuleTester = require('eslint').RuleTester;
+
+const { ERROR_MESSAGE } = rule;
+
+// ------------------------------------------------------------------------------
+// Tests
+// ------------------------------------------------------------------------------
+
+const ruleTester = new RuleTester({
+  parser: require.resolve('@babel/eslint-parser'),
+  parserOptions: { ecmaVersion: 2020, sourceType: 'module' },
+});
+
+ruleTester.run('no-array-prototype-extensions', rule, {
+  valid: [
+    '[1, 2, 4].filter(el => { /* ... */ })',
+    'something.filter(el => { /* ... */ })',
+    "filterBy(something, 'foo')",
+    'this.filterBy()',
+    'something[0]',
+    'something[0].foo',
+    "get(something, 'foo.0.bar')",
+    "get(something, 'foo.0')",
+    "get(something, 'foo')[0]",
+    "get(something, 'firstObjects')",
+    "get(something, 'lastObjects')",
+    'something.length',
+    'something[something.length - 1]',
+    /** Replace is part of string native prototypes */
+    "'something'.replace()",
+    'something.replace()',
+    'something.isAny',
+    "something['compact']",
+    /** Allow directly using Ember A() */
+    `
+     import { A } from '@ember/array';
+     something = A();
+     something.filterBy('foo');
+    `,
+    `
+      import { isArray } from '@ember/array';
+      something = isArray([1,2]);
+    `,
+  ],
+  invalid: [
+    {
+      code: '[1, 2, 3].filterBy()',
+      output: null,
+      errors: [{ message: ERROR_MESSAGE }],
+    },
+    {
+      code: 'something.filterBy()',
+      output: null,
+      errors: [{ message: ERROR_MESSAGE }],
+    },
+    {
+      code: 'getSomething().filterBy()',
+      output: null,
+      errors: [{ message: ERROR_MESSAGE }],
+    },
+    {
+      code: 'something.firstObject',
+      output: null,
+      errors: [{ message: ERROR_MESSAGE }],
+    },
+    {
+      code: "get(something, 'foo.firstObject')",
+      output: null,
+      errors: [{ message: ERROR_MESSAGE }],
+    },
+    {
+      code: "get(something, 'firstObject')",
+      output: null,
+      errors: [{ message: ERROR_MESSAGE }],
+    },
+    {
+      code: "get(something, 'foo.firstObject.bar')",
+      output: null,
+      errors: [{ message: ERROR_MESSAGE }],
+    },
+
+    {
+      code: "get(something, 'foo.lastObject')",
+      output: null,
+      errors: [{ message: ERROR_MESSAGE }],
+    },
+    {
+      code: "get(something, 'lastObject')",
+      output: null,
+      errors: [{ message: ERROR_MESSAGE }],
+    },
+    {
+      code: "get(something, 'foo.lastObject.bar')",
+      output: null,
+      errors: [{ message: ERROR_MESSAGE }],
+    },
+    {
+      code: 'something.firstObject()',
```

Comment:
I thought the string had to end with `firstObject` or end with `firstObject.` to be detected? I'm surprised it was detected with `()`.

Reply:
That’s because the `MemberExpression`-> firstObject structure for `something.firstObject` and `something.firstObject()` are exactly the same. In theory users should be able to do `.firstObject()`.  Added parent check to bypass in this case.

Reply:
I see, didn't know users could call it as a function, makes sense to detect it then.